src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/ModelBinderFactory.cs

```diff
@@ -122,26 +138,47 @@ private IModelBinder CreateBinderCore(DefaultModelBinderProviderContext provider
                 }
             }
 
-            if (result == null && stack.Count > 1)
+            // If the DelegatingBinder was created, then it means we recursed. Hook it up to the 'real' binder.
+            var delegatingBinder = collection[key] as PlaceholderBinder;
+            if (delegatingBinder != null)
             {
-                // Use a no-op binder if we're below the top level. At the top level, we throw.
-                result = NoOpBinder.Instance;
+                // It's also possible that user code called into `CreateBinder` but then returned null, we don't
+                // want to create something that will null-ref later so just hook this up to the no-op binder.
+                delegatingBinder.Inner = result ?? NoOpBinder.Instance;
```

Comment:
Wierd case. Now that we cache everything it would be possible for this to pollute the cache so we need to fix it. I wrote a test for it.